api/src/main/java/org/pmiops/workbench/billing/FreeTierBillingService.java

```diff
@@ -468,87 +372,68 @@ private void updateFreeTierUsage(
-        liveCostByCreator.keySet().stream()
-            .filter(
-                u ->
-                    CostComparisonUtils.compareCosts(
-                            dbCostByCreator.get(u), liveCostByCreator.get(u))
-                        != 0)
-            .collect(Collectors.toSet());
-
-    logger.info(String.format("Found %d users with changed costs.", usersWithChangedCosts.size()));
-
-    final Map<Long, DbUser> dbUsersWithChangedCosts = new HashMap<>();
-    userDao
-        .findAllById(usersWithChangedCosts)
-        .forEach(user -> dbUsersWithChangedCosts.put(user.getUserId(), user));
-    return dbUsersWithChangedCosts;
+  private boolean costAboveLimit(final DbUser user, final double currentCost) {
+    return CostComparisonUtils.costAboveLimit(
+        user, currentCost, workbenchConfigProvider.get().billing.defaultFreeCreditsDollarLimit);
   }
 
-  /**
-   * Get the list of newly expired users (who exceeded their free tier limit) and mark all their
-   * workspaces as inactive
-   *
-   * @param allUsers set of all users to filter them whether they have active free tier workspace
-   * @param dbUsersWithChangedCosts The map of users with changed costs, to calculate which of them
-   *     exceeded their free tier limit
-   * @param costByCreator Map of userId->dbCost
-   * @param liveCostByCreator Map of userId->liveCost
-   */
-  private void updateWorkspaceStatusToInactiveForNewlyExpiredUsers(
-      final Set<DbUser> allUsers,
-      Map<Long, DbUser> dbUsersWithChangedCosts,
-      Map<Long, Double> costByCreator,
-      Map<Long, Double> liveCostByCreator) {
-
-    Set<DbUser> freeTierUsers = getFreeTierActiveWorkspaceCreatorsIn(allUsers);
-
-    // Find users who exceeded their free tier limit
-    // Here costs in liveCostByCreator could be outdated because we're filtering on active  or
-    // recently deleted workspaces
-    // in previous steps. However, costByCreator will contain the up-to-date costs for all the other
-    // workspaces.
-    // This is why Math.max is used
-    final Set<DbUser> expiredUsers =
-        dbUsersWithChangedCosts.entrySet().stream()
-            .filter(
-                e ->
-                    costAboveLimit(
-                        e.getValue(),
-                        Math.max(costByCreator.get(e.getKey()), liveCostByCreator.get(e.getKey()))))
-            .map(Entry::getValue)
-            .collect(Collectors.toSet());
+  private Integer getMinutesBeforeLastFreeTierJob() {
+    return Optional.ofNullable(workbenchConfigProvider.get().billing.minutesBeforeLastFreeTierJob)
+        .orElse(120);
+  }
 
-    final Set<DbUser> newlyExpiredFreeTierUsers = Sets.intersection(expiredUsers, freeTierUsers);
+  @Nullable
+  private Map<String, Long> getWorkspaceByProjectCache(
+      List<WorkspaceCostView> allCostsInDbForUsers) {
+    // No need to proceed since there's nothing to update anyway
+    if (allCostsInDbForUsers.isEmpty()) {
+      return null;
```

Comment:
nit, we should avoid using a nullable collection, can we use empty map instead?  
[go/java-practices/null#avoid](go/java-practices/null#avoid)